LLVM crashes on diag function

The following program:
```effekt
type MyBoxedInt{ MkMyBoxedInt(value: Int) }
type Tuple2{
    Tuple2(l: MyBoxedInt, r: MyBoxedInt)
}

def diag(a: MyBoxedInt) = (a,a)

def main() = {
    val p = diag(MkMyBoxedInt(2))
    println(0)
}
```
[code file](https://gist.github.com/marzipankaiser/4f1f50fbfe6154cd5b90b950904418b5)

crashes using the LLVM backend, with the following error message:
```
opt: ./out/issue.ll:905:9: error: multiple definition of local value named 'a_102'
        %a_102 = load %Pos, %Pos* %a_102p.46
        ^
Exception in thread "main" java.lang.RuntimeException: Nonzero exit value: 1
```
[stack trace omitted]

--- a/effekt/shared/src/main/scala/effekt/generator/llvm/Transformer.scala
+++ b/effekt/shared/src/main/scala/effekt/generator/llvm/Transformer.scala
@@ -1,5 +1,6 @@
 package effekt.llvm
 
+import scala.collection.mutable
 import effekt.llvm.Operand.LocalReference
 import effekt.machine
 import effekt.machine.analysis.*
@@ -420,6 +421,24 @@ object Transformer {
     }
   }
 
+  def getEraser(environment: machine.Environment)(using C: ModuleContext): Operand = {
+    val types = environment.map{ _.tpe };
+    val freshEnvironment = environment.map{
+      case machine.Variable(name, tpe) => machine.Variable(freshName(name), tpe)
+    };
+    val eraser = ConstantGlobal(eraserType, freshName("eraser"));
+
+    C.erasers.getOrElseUpdate(types, {
+      defineFunction(eraser.name, List(Parameter(envType, "env"))) {
+        // TODO avoid unnecessary loads
+        loadEnvironmentAt(LocalReference(envType, "env"), freshEnvironment);
+        eraseValues(freshEnvironment, Set());
+        RetVoid()
+      };
+      eraser
+    });
+  }
+
   def produceObject(environment: machine.Environment, freeInBody: Set[machine.Variable])(using ModuleContext, FunctionContext, BlockContext): Operand = {
     if (environment.isEmpty) {
       ConstantNull(objType)
@@ -427,15 +446,7 @@ object Transformer {
       val obj = LocalReference(objType, freshName("obj"));
       val env = LocalReference(envType, freshName("env"));
       val size = ConstantInt(environmentSize(environment));
-      val eraser = ConstantGlobal(eraserType, freshName("eraser"));
-
-      //TODO cache eraser based on environment
-      defineFunction(eraser.name, List(Parameter(envType, "env"))) {
-        // TODO avoid unnecessary loads
-        loadEnvironmentAt(LocalReference(envType, "env"), environment);
-        eraseValues(environment, Set());
-        RetVoid()
-      }
+      val eraser = getEraser(environment)
 
       emit(Call(obj.name, objType, newObject, List(eraser, size)));
       emit(Call(env.name, envType, objectEnvironment, List(obj)));
@@ -653,6 +664,7 @@ object Transformer {
   class ModuleContext() {
     var counter = 0;
     var definitions: List[Definition] = List();
+    val erasers = mutable.HashMap[List[machine.Type], Operand]();
   }
 
   def emit(definition: Definition)(using C: ModuleContext) =

